File: android/app/src/test/java/com/clingsync/android/presentation/MainReducerTest.kt
package com.clingsync.android.presentation

import com.clingsync.android.AppSettings
import com.clingsync.android.FileStatus
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test
import java.io.File

class MainReducerTest {
    private val source = "/sdcard/DCIM"

    private fun settings(
        host: String = "s3+http://server-a",
        prefix: String = "/phone/",
        sourceDir: String = source,
        mediaOnly: Boolean = true,
    ) = AppSettings(
        hostUrl = host,
        repoPathPrefix = prefix,
        author = "Tester",
        sourceDirectory = sourceDir,
        mediaOnly = mediaOnly,
    )

    private fun path(name: String) = "$source/$name"

    private fun stateWithFiles(statuses: Map<String, FileStatus>) =
        MainUiState(
            settings = settings(),
            files = statuses.keys.map { File(it) },
            fileStatus = statuses,
            isConnected = true,
        )

    @Test
    fun toggleSelectionAddsThenRemoves() {
        val s0 = stateWithFiles(mapOf(path("a.jpg") to FileStatus.New))
        val selected = MainReducer.reduce(s0, MainEvent.FileSelectionChanged(path("a.jpg"), true)).state
        assertEquals(setOf(path("a.jpg")), selected.selectedPaths)
        val deselected = MainReducer.reduce(selected, MainEvent.FileSelectionChanged(path("a.jpg"), false)).state
        assertTrue(deselected.selectedPaths.isEmpty())
    }

    @Test
    fun selectAllPicksOnlySelectableFiles() {
        val s0 =
            stateWithFiles(
                mapOf(
                    path("new.jpg") to FileStatus.New,
                    path("synced.jpg") to FileStatus.Exists("phone/synced.jpg"),
                    path("done.jpg") to FileStatus.Done,
                    path("aborted.jpg") to FileStatus.Aborted,
                ),
            )

        val next = MainReducer.reduce(s0, MainEvent.SelectAllClicked).state

        assertEquals(setOf(path("new.jpg"), path("aborted.jpg")), next.selectedPaths)
    }

    @Test
    fun searchQueryChangeClearsSelection() {
        val s0 = stateWithFiles(mapOf(path("a.jpg") to FileStatus.New)).copy(selectedPaths = setOf(path("a.jpg")))
        val next = MainReducer.reduce(s0, MainEvent.SearchQueryChanged("foo")).state
        assertEquals("foo", next.searchQuery)
        assertTrue(next.selectedPaths.isEmpty())
    }

    @Test
    fun clearingSearchKeepsSelection() {
        val s0 =
            stateWithFiles(mapOf(path("a.jpg") to FileStatus.New))
                .copy(searchQuery = "a", selectedPaths = setOf(path("a.jpg")))

        val next = MainReducer.reduce(s0, MainEvent.SearchCleared).state

        assertEquals("", next.searchQuery)
        // Unlike editing the query, the clear-X preserves the selection.
        assertEquals(setOf(path("a.jpg")), next.selectedPaths)
    }

    @Test
    fun searchFilterMatchesFolderAndName() {
        val s0 =
            stateWithFiles(
                mapOf(
                    path("beach.jpg") to FileStatus.New,
                    "$source/vacation/sunset.jpg" to FileStatus.New,
                ),
            ).copy(searchQuery = "vacation")

        assertEquals(listOf(File("$source/vacation/sunset.jpg")), s0.displayedFiles)
    }

    @Test
    fun openingSearchClearsSelectionClosingClearsQuery() {
        val s0 = stateWithFiles(mapOf(path("a.jpg") to FileStatus.New)).copy(selectedPaths = setOf(path("a.jpg")))

        val opened = MainReducer.reduce(s0, MainEvent.SearchToggled).state
        assertTrue(opened.showSearch)
        assertTrue(opened.selectedPaths.isEmpty())

        val withQuery = opened.copy(searchQuery = "x")
        val closed = MainReducer.reduce(withQuery, MainEvent.SearchToggled).state
        assertFalse(closed.showSearch)
        assertEquals("", closed.searchQuery)
    }

    @Test
    fun uploadEnqueuesSelectedFilesAndMarksThemWaiting() {
        val s0 =
            stateWithFiles(mapOf(path("a.jpg") to FileStatus.New, path("b.jpg") to FileStatus.New))
                .copy(selectedPaths = setOf(path("a.jpg"), path("b.jpg")))

        val reduction = MainReducer.reduce(s0, MainEvent.UploadClicked)

        assertTrue(reduction.state.isUploadInitiated)
        assertEquals(setOf(path("a.jpg"), path("b.jpg")), reduction.state.currentUploadPaths)
        assertEquals(FileStatus.Waiting, reduction.state.fileStatus[path("a.jpg")])
        assertTrue(reduction.state.selectedPaths.isEmpty())
        assertEquals(2, reduction.state.uploadInfo?.totalFiles)
        assertEquals(
            listOf(Effect.EnqueueUpload(listOf(path("a.jpg"), path("b.jpg")), "Tester")),
            reduction.effects,
        )
    }

    @Test
    fun uploadWithNoSelectionDoesNothing() {
        val s0 = stateWithFiles(mapOf(path("a.jpg") to FileStatus.New))
        val reduction = MainReducer.reduce(s0, MainEvent.UploadClicked)
        assertEquals(s0, reduction.state)
        assertTrue(reduction.effects.isEmpty())
    }

    @Test
    fun abortResetsInitiatedAndRequestsCancel() {
        val s0 = stateWithFiles(emptyMap()).copy(isUploadInitiated = true)
        val reduction = MainReducer.reduce(s0, MainEvent.AbortClicked)
        assertFalse(reduction.state.isUploadInitiated)
        assertEquals(listOf(Effect.CancelUpload), reduction.effects)
    }

    @Test
    fun savingANewRepositoryInvalidatesTheOldOne() {
        val s0 =
            stateWithFiles(mapOf(path("a.jpg") to FileStatus.Done))
                .copy(selectedPaths = setOf(path("a.jpg")), isConnected = true)
        val newSettings = settings(host = "s3+http://server-b")

        val reduction = MainReducer.reduce(s0, MainEvent.SettingsSaved(newSettings))

        assertFalse(reduction.state.isConnected)
        assertTrue(reduction.state.fileStatus.isEmpty())
        assertTrue(reduction.state.selectedPaths.isEmpty())
        assertFalse(reduction.state.showSettings)
        assertEquals(
            listOf(
                Effect.PersistSettings(newSettings),
                Effect.InvalidateRepository("s3+http://server-a"),
                Effect.LoadFiles,
            ),
            reduction.effects,
        )
    }

    @Test
    fun changingOnlyTheSourceKeepsTheConnectionButReloads() {
        val s0 = stateWithFiles(mapOf(path("a.jpg") to FileStatus.Done)).copy(isConnected = true)
        val newSettings = settings(sourceDir = "/sdcard/Pictures")

        val reduction = MainReducer.reduce(s0, MainEvent.SettingsSaved(newSettings))

        assertTrue(reduction.state.isConnected)
        assertTrue(reduction.state.fileStatus.isEmpty())
        assertEquals(listOf(Effect.PersistSettings(newSettings), Effect.LoadFiles), reduction.effects)
    }

    @Test
    fun savingUnchangedSettingsJustPersists() {
        val s0 = stateWithFiles(mapOf(path("a.jpg") to FileStatus.Done)).copy(isConnected = true)

        val reduction = MainReducer.reduce(s0, MainEvent.SettingsSaved(s0.settings))

        assertTrue(reduction.state.isConnected)
        assertEquals(mapOf(path("a.jpg") to FileStatus.Done), reduction.state.fileStatus)
        assertEquals(listOf(Effect.PersistSettings(s0.settings)), reduction.effects)
    }

    @Test
    fun savingAnInvalidHostUrlShowsErrorAndDoesNotPersist() {
        val s0 = stateWithFiles(emptyMap())
        val reduction = MainReducer.reduce(s0, MainEvent.SettingsSaved(settings(host = "https://nope")))

        assertTrue(reduction.state.overlay is Overlay.Error)
        assertEquals("Invalid Host URL", (reduction.state.overlay as Overlay.Error).title)
        assertEquals(s0.settings, reduction.state.settings)
        assertTrue(reduction.effects.isEmpty())
    }

    @Test
    fun connectFailureShowsConnectionErrorWhenNoOverlayPresent() {
        val s0 = stateWithFiles(emptyMap()).copy(isConnecting = true)
        val next = MainReducer.reduce(s0, MainEvent.ConnectFailed("Failed to connect: nope")).state
        assertFalse(next.isConnecting)
        assertFalse(next.isConnected)
        assertEquals(Overlay.Error("Connection Error", "Failed to connect: nope"), next.overlay)
    }

    @Test
    fun scanFailureRevertsScanningFilesToNew() {
        val s0 =
            stateWithFiles(mapOf(path("a.jpg") to FileStatus.Scanning, path("b.jpg") to FileStatus.Done))
        val next = MainReducer.reduce(s0, MainEvent.ScanFailed("io", listOf(path("a.jpg"), path("b.jpg")))).state
        assertEquals(FileStatus.New, next.fileStatus[path("a.jpg")])
        // Only files that were Scanning are reverted.
        assertEquals(FileStatus.Done, next.fileStatus[path("b.jpg")])
        assertFalse(next.isScanning)
    }
}

File: ios/ClingSyncTests/MainReducerTest.swift
import Foundation
import Testing

@testable import ClingSync

struct MainReducerTest {
    private func config(
        host: String = "s3+http://server-a",
        prefix: String = "/phone/",
        author: String = "Tester"
    ) -> RepositoryConfiguration {
        RepositoryConfiguration(hostURL: host, repoPathPrefix: prefix, author: author)
    }

    private func file(_ name: String) -> SourceFile {
        SourceFile(id: name, name: name, size: 10, modificationDate: .distantPast)
    }

    private func stateWithFiles(_ statuses: [String: FileStatus]) -> AppState {
        var state = AppState(configuration: config())
        state.files = statuses.keys.sorted().map { file($0) }
        state.fileStatus = statuses
        state.isConnected = true
        return state
    }

    @Test func toggleSelectionAddsThenRemoves() {
        let state = stateWithFiles(["a.jpg": .new])
        let selected = MainReducer.reduce(state, .fileSelectionChanged(id: "a.jpg", selected: true)).state
        #expect(selected.selectedIds == ["a.jpg"])
        let deselected = MainReducer.reduce(selected, .fileSelectionChanged(id: "a.jpg", selected: false)).state
        #expect(deselected.selectedIds.isEmpty)
    }

    @Test func selectAllPicksOnlySelectableFiles() {
        let state = stateWithFiles([
            "new.jpg": .new,
            "synced.jpg": .exists(repoPath: "phone/synced.jpg"),
            "done.jpg": .done,
            "aborted.jpg": .aborted,
            "failed.jpg": .failed(message: "boom"),
        ])

        let next = MainReducer.reduce(state, .selectAllClicked).state

        #expect(next.selectedIds == ["new.jpg", "aborted.jpg", "failed.jpg"])
    }

    @Test func searchQueryChangeClearsSelection() {
        var state = stateWithFiles(["a.jpg": .new])
        state.selectedIds = ["a.jpg"]
        let next = MainReducer.reduce(state, .searchQueryChanged("foo")).state
        #expect(next.searchQuery == "foo")
        #expect(next.selectedIds.isEmpty)
    }

    @Test func clearingSearchKeepsSelection() {
        var state = stateWithFiles(["a.jpg": .new])
        state.searchQuery = "a"
        state.selectedIds = ["a.jpg"]

        let next = MainReducer.reduce(state, .searchCleared).state

        #expect(next.searchQuery == "")
        // Unlike editing the query, the clear-X preserves the selection.
        #expect(next.selectedIds == ["a.jpg"])
    }

    @Test func searchFilterMatchesName() {
        var state = stateWithFiles(["beach.jpg": .new, "sunset.jpg": .new])
        state.searchQuery = "sun"
        #expect(state.displayedFiles.map(\.id) == ["sunset.jpg"])
    }

    @Test func openingSearchClearsSelectionClosingClearsQuery() {
        var state = stateWithFiles(["a.jpg": .new])
        state.selectedIds = ["a.jpg"]

        let opened = MainReducer.reduce(state, .searchToggled).state
        #expect(opened.showSearch)
        #expect(opened.selectedIds.isEmpty)

        var withQuery = opened
        withQuery.searchQuery = "x"
        let closed = MainReducer.reduce(withQuery, .searchToggled).state
        #expect(!closed.showSearch)
        #expect(closed.searchQuery == "")
    }

    @Test func uploadEnqueuesSelectedFilesAndMarksThemWaiting() {
        var state = stateWithFiles(["a.jpg": .new, "b.jpg": .new])
        state.selectedIds = ["a.jpg", "b.jpg"]

        let reduction = MainReducer.reduce(state, .uploadClicked)

        #expect(reduction.state.isUploadInitiated)
        #expect(reduction.state.currentUploadIds == ["a.jpg", "b.jpg"])
        #expect(reduction.state.fileStatus["a.jpg"] == .waiting)
        #expect(reduction.state.selectedIds.isEmpty)
        #expect(reduction.state.uploadInfo?.totalFiles == 2)
        #expect(reduction.effects == [.enqueueUpload(ids: ["a.jpg", "b.jpg"], author: "Tester")])
    }

    @Test func uploadWithNoSelectionDoesNothing() {
        let state = stateWithFiles(["a.jpg": .new])
        let reduction = MainReducer.reduce(state, .uploadClicked)
        #expect(reduction.state == state)
        #expect(reduction.effects.isEmpty)
    }

    @Test func abortResetsInitiatedAndRequestsCancel() {
        var state = stateWithFiles([:])
        state.isUploadInitiated = true
        let reduction = MainReducer.reduce(state, .abortClicked)
        #expect(!reduction.state.isUploadInitiated)
        #expect(reduction.effects == [.cancelUpload])
    }

    @Test func savingANewRepositoryInvalidatesTheOldOne() {
        var state = stateWithFiles(["a.jpg": .done])
        state.selectedIds = ["a.jpg"]
        let newConfig = config(host: "s3+http://server-b")

        let reduction = MainReducer.reduce(state, .settingsSaved(newConfig))

        #expect(!reduction.state.isConnected)
        #expect(reduction.state.fileStatus.isEmpty)
        #expect(reduction.state.selectedIds.isEmpty)
        #expect(!reduction.state.showSettings)
        #expect(
            reduction.effects == [
                .persistSettings(newConfig),
                .invalidateRepository(repositoryID: "s3+http://server-a"),
                .loadFiles,
            ])
    }

    @Test func changingNonRepoSettingsJustPersists() {
        let state = stateWithFiles(["a.jpg": .done])
        let newConfig = config(author: "Renamed")

        let reduction = MainReducer.reduce(state, .settingsSaved(newConfig))

        #expect(reduction.state.isConnected)
        #expect(reduction.state.fileStatus == ["a.jpg": .done])
        #expect(reduction.effects == [.persistSettings(newConfig)])
    }

    @Test func savingUnchangedSettingsJustPersists() {
        let state = stateWithFiles(["a.jpg": .done])
        let reduction = MainReducer.reduce(state, .settingsSaved(state.configuration))
        #expect(reduction.state.isConnected)
        #expect(reduction.state.fileStatus == ["a.jpg": .done])
        #expect(reduction.effects == [.persistSettings(state.configuration)])
    }

    @Test func savingAnInvalidHostUrlShowsErrorAndDoesNotPersist() {
        let state = stateWithFiles([:])
        let reduction = MainReducer.reduce(state, .settingsSaved(config(host: "https://nope")))

        if case .error(let title, _) = reduction.state.overlay {
            #expect(title == "Invalid Host URL")
        } else {
            Issue.record("expected an error overlay")
        }
        #expect(reduction.state.configuration == state.configuration)
        #expect(reduction.effects.isEmpty)
    }

    @Test func connectFailureLandsOnTheConnectionFailedScreen() {
        var state = stateWithFiles([:])
        state.isConnecting = true
        let next = MainReducer.reduce(state, .connectFailed("Failed to connect: nope")).state
        #expect(!next.isConnecting)
        #expect(!next.isConnected)
        #expect(next.phase == .connectionFailed("Failed to connect: nope"))
    }

    @Test func scanFailureRevertsScanningFilesToNew() {
        let state = stateWithFiles(["a.jpg": .checking, "b.jpg": .done])
        let next = MainReducer.reduce(state, .scanFailed(message: "io", ids: ["a.jpg", "b.jpg"])).state
        #expect(next.fileStatus["a.jpg"] == .new)
        // Only files that were checking are reverted.
        #expect(next.fileStatus["b.jpg"] == .done)
        #expect(!next.isScanning)
        #expect(next.overlay == .error(title: "File Scanning Error", message: "Some files could not be scanned: io"))
    }
}
